packages/lodestar/src/eth1/impl/ethers.ts

```diff
@@ -63,19 +70,53 @@ export class EthersEth1Notifier extends (EventEmitter as { new(): Eth1EventEmitt
     this.logger.info(
       `Started listening on eth1 events on chain ${(await this.provider.getNetwork()).chainId}`
     );
+  }
 
-
+  public async initBlockCache(config: IBeaconConfig, state: BeaconState): Promise<void> {
+    const head = await this.getHead();
+    const range = getEth1BlockCandidateRange(config, state, head);
+    const promises: Promise<Block>[] = [];
+    for(let blockNumber = range.fromNumber; blockNumber < range.toNumber; blockNumber++) {
+      promises.push(this.getBlock(blockNumber));
+    }
+    const blocks = await Promise.all(promises);
+    this.blocksCache.init(blocks, head);
   }
 
   public async stop(): Promise<void> {
     this.provider.removeAllListeners("block");
     this.contract.removeAllListeners("DepositEvent");
   }
 
+  public pruneBlockCache(config: IBeaconConfig, finalizedState: BeaconState): void {
+    const timestamp = getLatestEth1BlockTimestamp(config, finalizedState);
+    this.blocksCache.prune(timestamp);
+  }
+
+  public findBLocks(config: IBeaconConfig, periodStart: Number64): Block[] {
+    const allBlocks = this.blocksCache.findBlocksByTimestamp();
+    return allBlocks.filter(block => isCandidateBlock(config, block, periodStart));
+  }
+
   public async processBlockHeadUpdate(blockNumber: number): Promise<void> {
     this.logger.verbose(`Received eth1 block ${blockNumber}`);
     const block = await this.provider.getBlock(blockNumber);
     this.emit("block", block);
+    let requestedBlockNumber = this.blocksCache.requestNewBlock(block);
+    if (requestedBlockNumber) {
+      let requestedBlock: Block;
+      let retry = 0;
+      while (!requestedBlock && retry < GET_ETH1_BLOCK_RETRY) {
+        requestedBlock = await this.getBlock(requestedBlockNumber);
+        requestedBlockNumber++;
+        retry++;
+      }
```

Comment:
We are doing this in bunch of places now. Can you create some utility function like:
```javascript
export async function retryable<T>(func: (...args) => T|null , retry = 3): T|null;
```
so we would have:
```javascript
const requestedBlock =  retryable(this.getBlock.bind(this, requestedBlockNumber), GET_ETH1_BLOCK_RETRY)
```

Reply:
I have to increase `requestedBlockNumber` on every call so have to come up with something like this which is quite complex:
```
export async function retryable<T>(funcGen: Generator<(...args: any[]) => Promise<T|null>>, retry = 3): Promise<T|null> {
  let result: T;
  let count = 0;
  while(!result && count < retry) {
    const func = funcGen.next().value;
    result = await func();
    count++;
  }
  return result;
}
```
And the consumers have to implement a generator. Any simpler approach?

Reply:
I think you should support plain function as well and check inside `retryable` if argument is iterator or plain function